### PR TITLE
Config sweeps

### PR DESCRIPTION
**Pull request by alvestrand**
This should be inspected primarily by @louquillio for sane HTML / Jekyll usage.

Python should be inspected by @phoglund.

Javascript should be inspected by someone....

### CODE REVIEW

File: bin/generate_sweep_data [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import encoder
+import glob
+import json
+import optimizer
+import pick_codec
+import re
+import score_tools
+import sys
+
+def ReportOneSweep(codec_name, criterion_name, videofile_name, encoder_name):
```
Comment (phoglund) [inline bin/generate_sweep_data]: Nit: two blank lines

File: bin/generate_sweep_data [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  sweepdata = {}
+  codecs = {}
+  for filename in files:
+    with open(filename, 'r') as file:
+      resultobject = json.load(file)
+      if not 'detailed' in resultobject:
+        # Not a details file. Skip.
+        continue
+      # Extract criterion from filename. Not really pretty.
+      m = re.search(r'-([a-z]+)\.json$', filename)
```
Comment (phoglund) [inline bin/generate_sweep_data]: I suggest you document in the file header how one annotates in criteria into filenames (and provide an example).

File: bin/generate_sweep_data [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      details = resultobject['detailed']
+      # Record codec information. Should be consistent across files.
+      codecs.update(resultobject['codecs'])
+      for codec_name in details.keys():
+        for videofilename in details[codec_name].keys():
+          encoders = set()
+          for encoding in details[codec_name][videofilename]:
+            encoders.add(encoding['config_id'])
+          for encoder in encoders:
+            encodinglist = sweepdata.setdefault(codec_name, {}) \
```
Comment (phoglund) [inline bin/generate_sweep_data]: You never read the encodinglist variable; what is it good for?

File: bin/generate_sweep_data [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                .setdefault(videofilename, {})
```
Comment (phoglund) [inline bin/generate_sweep_data]: Oh, I see, you get a mutable entry out of the sweepdata dict.

Please don't chain setdefault calls. Rewrite this to
codec = sweepdata.setdefault(codec_name, {})
encodinglist = sweepdata.setdefault(videofilename, {})

It also seems unclear to me why a videofilename maps to an encoding list. Could the name be better, like encodings_for_video? Furthermore, I just guessed that codec_name maps to data for a codec, so could that be called sweeps_for_codec or something?

File: bin/generate_sweep_data [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            encodinglist[encoder] = (
```
Comment (phoglund) [inline bin/generate_sweep_data]: Also, you are four for loops deep here. Consider extracing a method on some suitable level to flatten this out a bit. For instance, GenerateSweepsForCodec(codec_name, ...), at least if it doesn't require an excessive amount of parameters.

File: lib/encoder.py [hunk 6]
```diff
@@ -810,7 +812,11 @@ def ReadEncodingResult(self, encoding, scoredir=None):
     if os.path.isfile(filename):
       with open(filename, 'r') as resultfile:
         stringbuffer = resultfile.read()
-        return ast.literal_eval(stringbuffer)
+        try:
+          return ast.literal_eval(stringbuffer)
```
Comment (phoglund) [inline lib/encoder.py]: This looks familiar; rebase problem=

**Comment by phoglund** [comment]
I had some comments on the python code; please have a look.

File: bin/generate_sweep_data (hunk 1 above)
Comment (alvestrand) [inline bin/generate_sweep_data]: done

File: bin/generate_sweep_data (hunk 2 above)
Comment (alvestrand) [inline bin/generate_sweep_data]: Done.

File: bin/generate_sweep_data (hunk 4 above)
Comment (alvestrand) [inline bin/generate_sweep_data]: Added a comment trying to explain. Also more lines for the setdefaults.

**Comment by phoglund** [comment]
lgtm

File: bin/generate_sweep_data (hunk 5 above)
Comment (alvestrand) [inline bin/generate_sweep_data]: Three levels; the two last for loops are on the same level.

Does this look better?

File: lib/encoder.py (hunk 6 above)
Comment (alvestrand) [inline lib/encoder.py]: Fixed. git diff --stat now shows no change to encoder.py.

File: bin/generate_sweep_data (hunk 5 above)
Comment (phoglund) [inline bin/generate_sweep_data]: Yeah, that's better.

/ P

**Pull request merged**